### PR TITLE
Upgrade to swagger2

### PR DESCRIPTION
**Pull request by jbarroso**
Hello!

This is an attempt to upgrade this module to Swagger UI 2.2.6  / OpenApi 2.0

The main changes are:
- Javascript library + assets upgraded.
- Unit Tests with fixtures added. Now we can check the expected result.
- Model definitions are completed using Request and Response documentation example data. With this approach we can have rich model definitions that have been taken from the Apigility's Admin UI..

Hope it is ok! Everything that I have tested is working and we are using it in production for one of our API's

Best regards!
Jose.

### CODE REVIEW

**Comment by ikovalyov** [comment]
any update here?

File: src/Service.php [hunk 1]
```diff
@@ -7,179 +8,315 @@
...
      * @var BaseService
      */
     protected $service;
+    protected $modelGenerator;
 
     /**
      * @param BaseService $service
-     * @param string $baseUrl
      */
-    public function __construct(BaseService $service, $baseUrl)
```
Comment (weierophinney) [inline src/Service.php]: Why do you remove the `$baseUrl` parameter? This is necessary when hosting the service under a separate directory of a primary website...

File: src/Service.php [hunk 2]
```diff
@@ -7,179 +8,315 @@
...
-        // localize service object for brevity
-        $service = $this->service;
+        return $this->cleanEmptyValues([
+            'tags' => $this->getTags(),
+            'paths' => $this->cleanEmptyValues($this->getPaths()),
+            'definitions' => $this->getDefinitions()
+        ]);
+    }
+
+    protected function getTags()
```
Comment (weierophinney) [inline src/Service.php]: We favor `private` visibility over `protected` whenever possible. `protected` is part of the public API, and is thus problematic if we want to later change the signature or remove the method.

File: src/Service.php [hunk 3]
```diff
@@ -7,179 +8,315 @@
...
+    }
+
+    protected function getPathOperation(Operation $operation, $parameters)
+    {
+        return $this->cleanEmptyValues([
+                'tags' => [$this->service->getName()],
+                'description' => $operation->getDescription(),
+                'parameters' => $parameters,
+                'produces' => $this->service->getRequestAcceptTypes(),
+                'responses' => $this->getResponsesFromOperation($operation)
```
Comment (weierophinney) [inline src/Service.php]: Please reduce the level of indentation on these lines by 1 (4 spaces).

File: src/Service.php [hunk 4]
```diff
@@ -7,179 +8,315 @@
...
+    }
+
+    protected function serviceContainsPostPutOrPatchMethod()
+    {
+        foreach ($this->getAllOperations() as $operation) {
+            $method = $this->getMethodFromOperation($operation);
+            if ($this->isMethodPostPutOrPatch($method)) {
+                return true;
             }
+        }
```
Comment (weierophinney) [inline src/Service.php]: Method should return `false` at the end, no?

File: src/Service.php [hunk 5]
```diff
@@ -7,179 +8,315 @@
...
+            ]);
+        }
+        return $responses;
+    }
+
+    protected function getResponseSchema(Operation $operation, $code)
+    {
+        if ($code === 200 || $code === 201) {
+            return $this->modelGenerator->generate($operation->getResponseDescription());
+        }
```
Comment (weierophinney) [inline src/Service.php]: Is it expected to have a `void`/`null` return type for non-200/201 response types here?

File: src/Service.php [hunk 6]
```diff
@@ -7,179 +8,315 @@
...
+        foreach ($this->getAllOperations() as $operation) {
+            $method = $this->getMethodFromOperation($operation);
+            if ($method === 'post') {
+                return $operation->getRequestDescription();
             }
-            $operationGroups[] = [
-                'operations' => $operations,
-                'path'       => $routeWithReplacements
-            ];
         }
```
Comment (weierophinney) [inline src/Service.php]: Is this method expected to return `void`/`null` if no `post` operations are found?

File: src/Service.php [hunk 7]
```diff
@@ -7,179 +8,315 @@
...
-                    'required'   => $requiredProperties,
-                    'properties' => $properties,
-                ],
-            ],
-        ];
+    protected function getFieldType(Field $field)
+    {
+        return (method_exists($field, 'getFieldType') &&
+            !empty($field->getFieldType())) ?
+            $field->getFieldType() : self::DEFAULT_TYPE;
```
Comment (weierophinney) [inline src/Service.php]: Re-flow this to have each line have a single semantic:

```php
return method_exists($field, 'getFieldType') && ! empty($field->getFieldType())
    ? $field->getFieldType()
    : self::DEFAULT_TYPE;
```

This makes it easier to understand that the line is a ternary condition, and breaks it into the condition, the condition-matched return, and the unmatched return.

File: src/Service.php (hunk 1 above)
Comment (weierophinney) [reply src/Service.php]: It appears it was never used.

Comment (weierophinney) [reply src/Service.php]: Never mind; it was. The `ViewModel` class was creating `Service` instances previously, providing the baseUrl as a parameter, which was then used in the `basePath` element of the Swagger documentation. However, it appears this is no longer an element, which is why we can remove it here.

**Comment by weierophinney** [comment]
Thanks, @jbarroso!

**Pull request merged**